File: backend/ctgov.ts
import type { QueryResult } from "./cache";

export const CT_GOV_BASE = "https://clinicaltrials.gov/api/v2/studies";
const PAGE_CAP = 20; // max 20,000 studies per request

export async function fetchAllPages(baseParams: URLSearchParams): Promise<QueryResult> {
  baseParams.set("pageSize", "1000");
  baseParams.delete("pageToken");

  const allStudies: unknown[] = [];
  let pageToken: string | undefined;
  let page = 0;

  do {
    page++;
    const params = new URLSearchParams(baseParams);
    if (pageToken) params.set("pageToken", pageToken);

    const url = `${CT_GOV_BASE}?${params.toString()}`;
    console.log(`[/api/trials/all] page ${page} — ${url}`);

    const response = await fetch(url, { headers: { Accept: "application/json" } });
    if (!response.ok) throw new Error(`CT.gov returned ${response.status}`);

    const data = await response.json() as { studies?: unknown[]; nextPageToken?: string };
    allStudies.push(...(data.studies ?? []));
    pageToken = data.nextPageToken;

    console.log(
      `[/api/trials/all] page ${page}: ${(data.studies ?? []).length} studies (running total: ${allStudies.length})`
    );
  } while (pageToken && page < PAGE_CAP);

  if (pageToken && page >= PAGE_CAP) {
    console.warn(`[/api/trials/all] hit ${PAGE_CAP}-page cap — results may be incomplete`);
  }

  return { studies: allStudies, totalCount: allStudies.length };
}

File: backend_py/ctgov.py
import hashlib
import time
import requests

CT_GOV_BASE = "https://clinicaltrials.gov/api/v2/studies"
PAGE_CAP = 32  # max 32,000 studies per request


def fetch_all_pages(params: dict) -> list[dict]:
    """Fetch all paginated results from the CT.gov API for a given query."""
    params = {**params, "pageSize": 1000}
    params.pop("pageToken", None)

    all_studies = []
    page_token = None
    page = 0

    while True:
        page += 1
        req_params = {**params}
        if page_token:
            req_params["pageToken"] = page_token

        print(f"[ctgov] page {page} — params: {req_params}")
        response = requests.get(CT_GOV_BASE, params=req_params, headers={"Accept": "application/json"})
        response.raise_for_status()

        data = response.json()
        studies = data.get("studies", [])
        all_studies.extend(studies)
        page_token = data.get("nextPageToken")

        print(f"[ctgov] page {page}: {len(studies)} studies (running total: {len(all_studies)})")

        if not page_token or page >= PAGE_CAP:
            break

        time.sleep(0.1)  # be polite to the API

    if page_token and page >= PAGE_CAP:
        print(f"[ctgov] WARNING: hit {PAGE_CAP}-page cap — results may be incomplete")

    return all_studies


def hash_params(params: dict) -> str:
    sorted_str = "&".join(f"{k}={v}" for k, v in sorted(params.items()))
    return hashlib.md5(sorted_str.encode()).hexdigest()
